src/agent/codecoveragepublisher.ts

```diff
@@ -0,0 +1,221 @@
+import testifm = require('vso-node-api/interfaces/TestInterfaces');
+import ctxm = require('./context');
+import cm = require('./common');
+import utilities = require('./utilities');
+import buildifm = require('vso-node-api/interfaces/BuildInterfaces');
+import fc = require('./filecontainerhelper');
+import Q = require('q');
+import fs = require('fs');
+
+var shell = require('shelljs');
+var path = require('path');
+
+export class CodeCoveragePublisher {
+
+    constructor(executionContext: cm.IExecutionContext, command: cm.ITaskCommand, reader: ICodeCoverageReader) {
+        this.executionContext = executionContext;
+        this.command = command;
+        this.codeCoverageReader = reader;
+        this.buildId = parseInt(this.executionContext.variables[ctxm.WellKnownVariables.buildId]);
+        this.project = this.executionContext.variables[ctxm.WellKnownVariables.projectId];
+    }
+
+    private executionContext: cm.IExecutionContext;
+
+    private command: cm.ITaskCommand;
+
+    private codeCoverageReader: ICodeCoverageReader;
+
+    private buildId: number;
+
+    private project: string;   
+    
+    //-----------------------------------------------------
+    // Publish code coverage
+    //-----------------------------------------------------
+    public publishCodeCoverageSummary(): Q.Promise<boolean> {
+        var defer = Q.defer<boolean>();
+        var _this = this;
+        var summaryFile = _this.command.properties["summaryfile"];
+
+        _this.readCodeCoverageSummary(summaryFile).then(function(codeCoverageData) {
+            if (codeCoverageData && codeCoverageData.coverageStats && codeCoverageData.coverageStats.length > 0) {
```

Comment:
unable to find code coverage file and code coverage file stats are invalid are throwing same error message. separate them